### PR TITLE
feat(blob): multipart upload

### LINKED ISSUE #66: Multipart API support
It would be great if we support the [r2 multipart api](https://developers.cloudflare.com/r2/api/workers/workers-multipart-usage/).

Workers have a [memory limit of 128MB per Worker](https://developers.cloudflare.com/workers/platform/limits#worker-limits) so the blob can't resolve some large files. And multipart upload can imporve the experense of user who using low speed network

looks not too complex and I already have some code, I can make a pr if you want

### PR DESCRIPTION
**Pull request by Teages**
Resolves #66 

This pull request added multipart upload api and its proxy route.

Also fixed a type error on [`proxyHubBlob`](https://github.com/nuxt-hub/core/blob/cb4738059d7a9b8014aa9a56333754fa9a7abdf3/src/runtime/server/utils/blob.ts#L277), it breaks the type check

did some check in playground, ~~not check the proxy api yet~~.

Checked in remote proxy, works well except the `contentType` not included in the complete response (but it returns in dev server), maybe it takes time for r2 to determine the type

### CODE REVIEW

**Review by farnabaz — COMMENTED** [review]
I was wondering how we can change API routes to remove query params.

Just an idea:
- `/blob/multipart/[...pathname].post.ts` create a multip part upload
- `/blob/multipart/[...pathname]/[uploadId]/[partNumber].put.ts` put a part
- `/blob/multipart/[...pathname]/[uploadId]/complete.post.ts` complete upload task

**Comment by farnabaz** [comment]
Although that upload seems fine in dev, I have this wired error in terminal.

Screenshot 2024-04-12 at 01 35 17

**Comment by farnabaz** [comment]
Although that upload works fine, I have this wired error in terminal

Screenshot 2024-04-12 at 17 05 48

**Comment by Teages** [comment]
> Although that upload works fine, I have this wired error in terminal
> 
> Screenshot 2024-04-12 at 17 05 48

it is from the wroked, as [their member said](https://github.com/cloudflare/workerd/issues/1472#issuecomment-1850322438): 
> This message is not really an error, at least not one that's relevant to most people. This is another example where our logging is highlighting bugs meant for the workers runtime team to look into. In this case V8's API for mapping JITed memory in order to construct stack traces isn't working the way we expect. This has been firing occasionally for a long time. There's actually no impact for workerd users since this information is only used when generating crash dumps in our internal codebase.

so we can safity ignore this

**Comment by Teages** [comment]
> I was wondering how we can change API routes to remove query params.
> 
> Just an idea:
> 
> * `/blob/multipart/[...pathname].post.ts` create a multip part upload
> * `/blob/multipart/[...pathname]/[uploadId]/[partNumber].put.ts` put a part
> * `/blob/multipart/[...pathname]/[uploadId]/complete.post.ts` complete upload task

sadily radix3 doesn't support this route
```ts
import { createRouter } from "radix3"

const router = createRouter()

router.insert('/api/blob/multipart/**:pathname/:uploadId/:partNumber', { payload: 'found' })

const res = router.lookup('/blob/multipart/hello.txt/id/1')

console.log(res) // null
```

so I  just make create / complete in different route

**Comment by Atinux** [comment]
Thank you so much @Teages 

I think to move forward, I would like to have a `multipart: true` option for `hubBlob().put()` as automatic multipart:

```ts
await hubBlob().put('movie.mp4', file, {
  multipart: true
})
```

As well as updating the documentation for it.

As a bonus, it would be nice to find a way to have a Vue composable as well:

```ts
useUpload(`/api/blob/multipart/${file.name}`).upload(file, { multipart: true })
```

**Comment by Teages** [comment]
> I think to move forward, I would like to have a `multipart: true` option for `hubBlob().put()` as automatic multipart:
> 
> ```ts
> await hubBlob().put('movie.mp4', file, {
>   multipart: true
> })
> ```
> 

I think it maybe not not be very useful, hubBlob runs on nitro, which is running in cloudflare worker, it should have the most stable connection to cloudflare r2.

The limitation of direct upload seems to be 300mb, but woker only have 128mb memory. I can't see any need for workers to upload to r2 in multipart.

It should be the same for other platforms.

> As a bonus, it would be nice to find a way to have a Vue composable as well:
> 
> ```ts
> useUpload(`/api/blob/multipart/${file.name}`).upload(file, { multipart: true })
> ```

Yes I think so, but it affects flexibility, I didn't see nuxt hub have provided any upload api except hub proxy.

Or we can provide a composable like this:
```ts
const { progress } = useMultipartUpload({
  file,
  partSize: 10 * 1024 * 1024, // 10MB
  throttle: 2, // 2 concurrent uploads
}, ({partNumber, body}) => $fetch())
```

**Comment by Atinux** [comment]
> I think it maybe not not be very useful, hubBlob runs on nitro, which is running in cloudflare worker, it should have the most stable connection to cloudflare r2.

Makes sense, let's skip it.

What about:

```ts
const { progress, isComplete } = useMultipartUpload(file, {
  partSize: 10 * 1024 * 1024, // 10MB
  throttle: 2, // 2 concurrent uploads
}, ({ partNumber, chunkBody }) => $fetch())
```

It would be nice to have the second parameter as optional:

```ts
const { progress, isComplete } = useMultipartUpload(file, ({ partNumber, chunkBody }) => $fetch())
```

So it could use defaults we can define, wdyt?

**Comment by Teages** [comment]
well I designed a composable creater:

```ts
export const useMultipartUpload = createMultipartUploader({
  create: (file) => $fetch(),
  upload: ({ partNumber, chunkBody }, { pathname, uploadId }) => $fetch(),
  complete: (parts, { pathname, uploadId }) => $fetch(),
  abort: ({ pathname, uploadId }) => $fetch(),
  partSize: 10 * 1024 * 1024, // 10MB
  concurrent: 2, // 2 concurrent uploads
  maxRetry: 3,
})
```
then upload file in just two lines:

```ts
const { completed, progress, cancel } = useMultipartUpload(file)
const result = await completed
```

File: docs/content/docs/2.storage/3.blob.md [hunk 1]
```diff
@@ -189,6 +189,150 @@ You can also use the `delete()` method as alias of `del()`.
...
+  const mpu = hubBlob().resumeMultipartUpload(pathname, uploadId)
+  await mpu.abort()
+
+  return sendNoContent(event)
+})
+```
+
+A simple example of multipart upload in client with above routes:
+
+```ts
```
Comment (Atinux) [inline docs/content/docs/2.storage/3.blob.md:280]:
```suggestion
```ts [utils/multipart-upload.ts]
```

File: src/runtime/composables/multipart.ts [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+  maxRetry?: number
+}
+
+export type MultipartUploader<
+  TCreateResponse,
+  TUploadResponse,
+  TCompleteResponse,
+> = (
+  /**
+   * The file to be uploaded.
```
Comment (Atinux) [inline src/runtime/composables/multipart.ts:185]:
```suggestion
   * The file to upload.
```

**Comment by Atinux** [comment]
Thank you so much for your work, we are getting close!

I've been thinking, what about using the `action` as first part?

```
/api/blob/multipart/[action]/[...pathname].ts
```

We could have an abstraction similar to this:

```ts
// server/api/blob/multipart/[action]/[...pathname].ts
export default eventHandler(async (event) => {
  await hubBlob().handleMultipartUpload(event, {
    // options...
  })
})
```

What do you think? This will avoid creating 4 files to handle multipart upload, and it should also simplify the createMultipartUploader` function as we can simplify specify the base path?

**Comment by Teages** [comment]
That would be great, what options do we need for `handleMultipartUpload`? I would like to provide some hooks like `beforeCreate`, `afterCompleted` and etc, but we can let developers do themself like
```ts
// server/api/blob/multipart/[action]/[...pathname].ts
export default eventHandler(async (event) => {
  const actions = getRouterParam(event, 'action')
  // do something before handle
  await hubBlob().handleMultipartUpload(event, {
    action: getRouterParam(event, 'my-action-key'), // default getRouterParam(event, 'action'),
    filename: getRouterParam(event, 'my-filename-key'), // default getRouterParam(event, 'filename'),
  })
})
```

**Comment by Atinux** [comment]
> I would like to provide some hooks like beforeCreate, afterCompleted and etc,

The goal of exposing a function is that they can do whatever they want before and after the function instead of exposing an API route :)

I like the `fetchOptions` property for `useMultipartUpload` 🚀

**Comment by Atinux** [comment]
You are very fast @Teages 🚀 

Happy to resolve the conflicts? 🙏

**Comment by Teages** [comment]
oops seems I messed up, I need some time to solve

**Comment by Teages** [comment]
ok it looks good now

**Comment by Teages** [comment]
I noticed #99, everything looks fine except the way to resolve returning is different.
In this pr I use `respondWith` to force the response to be the handled response, but in #99 it just return the response and lets the user return on their own.

Do I need to change to the same approach?

**Comment by Atinux** [comment]
cc @farnabaz

**Pull request merged**